flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adapter/ExecutionGraphToSchedulingTopologyAdapter.java

```diff
@@ -0,0 +1,97 @@
+import org.apache.flink.runtime.scheduler.strategy.ExecutionVertexID;
+import org.apache.flink.runtime.scheduler.strategy.SchedulingExecutionVertex;
+import org.apache.flink.runtime.scheduler.strategy.SchedulingResultPartition;
+import org.apache.flink.runtime.scheduler.strategy.SchedulingTopology;
+import org.apache.flink.util.Preconditions;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * Adapter of {@link ExecutionGraph} to {@link SchedulingTopology}.
+ */
+public class ExecutionGraphToSchedulingTopologyAdapter implements SchedulingTopology {
+
+	private final Map<ExecutionVertexID, DefaultExecutionVertex> executionVertices;
+
+	private final List<DefaultExecutionVertex> executionVerticesList;
+
+	private final Map<IntermediateResultPartitionID, DefaultResultPartition> resultPartitions;
+
+	public ExecutionGraphToSchedulingTopologyAdapter(ExecutionGraph graph) {
+		Preconditions.checkNotNull(graph, "execution graph can not be null");
+
+		this.executionVertices = new HashMap<>();
+		this.resultPartitions = new HashMap<>();
+		this.executionVerticesList = new ArrayList<>();
+
+		Map<ExecutionVertex, DefaultExecutionVertex> executionVertexMap = new HashMap<>();
+		for (ExecutionVertex vertex : graph.getAllExecutionVertices()) {
+			DefaultExecutionVertex scheduleVertex = new DefaultExecutionVertex(vertex);
+			executionVertices.put(scheduleVertex.getId(), scheduleVertex);
+			executionVerticesList.add(scheduleVertex);
+			executionVertexMap.put(vertex, scheduleVertex);
+
+			for (SchedulingResultPartition srp : scheduleVertex.getProducedResultPartitions()) {
+				DefaultResultPartition drp = (DefaultResultPartition) srp;
+				resultPartitions.put(drp.getId(), drp);
+			}
+		}
+
+		for (Map.Entry<ExecutionVertex, DefaultExecutionVertex> mapEntry : executionVertexMap.entrySet()) {
+			final DefaultExecutionVertex scheduleVertex = mapEntry.getValue();
+			final ExecutionVertex executionVertex = mapEntry.getKey();
+
+			for (int index = 0; index < executionVertex.getNumberOfInputs(); index++) {
+				for (ExecutionEdge edge : executionVertex.getInputEdges(index)) {
+					DefaultResultPartition partition = resultPartitions.get(edge.getSource().getPartitionId());
+					scheduleVertex.addConsumedPartition(partition);
+					partition.addConsumer(scheduleVertex);
+				}
+			}
+		}
+	}
+
+	@Override
+	public Iterable<SchedulingExecutionVertex> getVertices() {
+		return Collections.unmodifiableList(executionVerticesList);
+	}
+
+	@Override
+	public Optional<SchedulingExecutionVertex> getVertex(ExecutionVertexID executionVertexId) {
+		return Optional.of(executionVertices.get(executionVertexId));
+	}
+
+	@Override
+	public Optional<SchedulingResultPartition> getResultPartition(IntermediateResultPartitionID intermediateResultPartitionId) {
+		return Optional.of(resultPartitions.get(intermediateResultPartitionId));
```

Comment:
you likely want to use `Optional#ofNullable` here and in `#getVertex´, `Optional#of` fails with an NPE when applied to `null`.

Reply:
OK